src/mon/FSCommands.cc

```diff
@@ -1261,6 +1259,193 @@ class RenameFilesystemHandler : public FileSystemCommandHandler
+	return -EPERM;
+      } else if (fs1p->get_mds_map().get_num_up_mds() == 0 &&
+		 fs2p->get_mds_map().get_num_up_mds() > 0) {
+	ss << "CephFS \"" << fs2_name << "\" has not been failed. Before "
+	   << "swapping CephFS names, both CephFSs should be marked as "
+	   << "failed. See `ceph fs fail`.";
+	return -EPERM;
+      } else {
+	ss << "CephFSs \"" << fs1_name << "\" and \"" << fs2_name << "\" has "
+	   << "not been failed. Before swapping CephFS names, both CephFSs "
+	   << "should be marked as failed. See `ceph fs fail`.";
+	return -EPERM;
+      }
+    }
+
+    // Check that refuse_client_session is set.
+    if (!fs1p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION) ||
+	!fs2p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION)) {
+      if (!fs1p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION) &&
+          fs2p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION)) {
+	ss << "Flag \"refuse_client_session\" is not set for \""
+	   << fs1_name << "\". See `ceph fs set`.";
+	return -EPERM;
+      } else if (
+          fs1p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION) &&
+	  !fs2p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION)) {
+	ss << "Flag \"refuse_client_session\" is not set for \""
+	   << fs2_name << "\". See `ceph fs set`.";
+	return -EPERM;
+      } else if (
+          !fs1p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION) &&
+	  !fs2p->get_mds_map().test_flag(CEPH_MDSMAP_REFUSE_CLIENT_SESSION)) {
+	ss << "Flag \"refuse_client_session\" is not set for both the "
+	   << "CephFSs. See `ceph fs set`.";
+	return -EPERM;
+      }
+    }
+
+    // Finally, the swap begins.
+    // Swap CephFS names on OSD pool application tag
+    for (const auto p : fs1p->get_mds_map().get_data_pools()) {
+      mon->osdmon()->do_application_enable(p, APP_NAME_CEPHFS, "data",
+					   fs2_name, true);
+    }
+    mon->osdmon()->do_application_enable(
+      fs1p->get_mds_map().get_metadata_pool(), APP_NAME_CEPHFS, "metadata",
+      fs2_name, true);
+
+    for (const auto p : fs2p->get_mds_map().get_data_pools()) {
+      mon->osdmon()->do_application_enable(p, APP_NAME_CEPHFS, "data",
+					   fs1_name, true);
+    }
+    mon->osdmon()->do_application_enable(
+      fs2p->get_mds_map().get_metadata_pool(), APP_NAME_CEPHFS, "metadata",
+      fs1_name, true);
+    mon->osdmon()->propose_pending();
+
+    // Now swap CephFS names and, optionally, FSCIDs.
+    auto renamefunc1 = [fs2_name](auto&& fs) {
+      fs.get_mds_map().set_fs_name(fs2_name);
+    };
+    fsmap.modify_filesystem(fs1_id, std::move(renamefunc1));
+    auto renamefunc2 = [fs1_name](auto&& fs) {
+      fs.get_mds_map().set_fs_name(fs1_name);
+    };
+    fsmap.modify_filesystem(fs2_id, std::move(renamefunc2));
+
+    if (swap_fscids_flag == "yes") {
+      fsmap.swap_fscids(fs1_id, fs2_id);
+    }
+
+    ss << "File system names have been swapped; cephx credentials needs an "
+	  "upgrade.";
```

Comment:
So I think we need to flush this out in the docs better. The admin only needs to **update** cephx credentials if clients formerly mounted on `<fs1>` need to continue accessing that fs as name `<fs2>` (and vice versa). I think for the expected use-case of this API (and you should note this in the doc), the swap is for the purposes of disaster recovery so the admin probably **wants** clients to continue using the existing cephx credentials because the file system name is (still) correct.

Reply:
Since we've made the changes in docs, I am marking this as resolved.